[mlir][arith] fix wrong floordivsi fold (#83079)

lipracer opened the pull request:
Fixs https://github.com/llvm/llvm-project/issues/83079

kuhar reviewed (commented):
Looks fine upon a quick scan of the patch. Given the number or error code paths, I think it could be made more readable with something like `constFoldBinaryOpConditional` IIRC.

pingshiyu commented:
Not sure this is the right patch, please see [this comment](https://github.com/llvm/llvm-project/issues/83079#issuecomment-1969041763)

lipracer commented:
@kuhar There are indeed many overflow states that need to be checked here, but I haven't come up with a good method yet. Can you give me a specific suggestion? Thank you.

kuhar commented:
> @kuhar There are indeed many overflow states that need to be checked here, but I haven't come up with a good method yet. Can you give me a specific suggestion? Thank you.

Without knowing the exact details of the floordivsi algorithm, the way I'd think about it is that we only need to detect if the *final* overflow happened or not. And of course fold correctly when there is no overflow. If there are intermediate overflows, I'd think that they can fold into one of the two buckets: (a) those that indicate that the floordivsi overflows for these arguments, and (b) implementation bugs.

It might be helpful to extract this fold code to a unit test (say one of the APInt test files) and go from there. Pick a few inputs that are known to overflow and trace the values and intermediate overflows. Maybe write a separate check to decide if overflow happens for the given input instead of relying on some intermediate checks.

kuhar commented:
@lipracer Another idea: we could perform the division on APInt with more bits than the bitwidth of the folded integer type. Maybe this way it'd be easier to detect overflow. And then once we are confident there's no overflow, truncate it back to the desired bitwidth.

kuhar reviewed (changes requested):
@lipracer could you open a separate PR for the APInt change? This will make it easier to review and land properly.

lipracer commented:
Ok.

lipracer committed:
[mlir][arith] fix wrong floordivsi fold (#83079)

1) fix floordivsi error expand logic
2) fix floordivsi fold did't check overflow stat

Fixs https://github.com/llvm/llvm-project/issues/83079

lipracer commented:
submit a APInt change [PR](https://github.com/llvm/llvm-project/pull/84720).

dwblaikie commented on llvm/lib/Support/APInt.cpp:
```diff
@@ -2022,6 +2022,14 @@ APInt APInt::ushl_ov(unsigned ShAmt, bool &Overflow) const {
+  auto quotient = sdiv_ov(RHS, Overflow);
+  if ((quotient * RHS != *this) && (isNegative() != RHS.isNegative()))
+    return quotient - 1;
+  else
+    return quotient;
```
LLVM style encourages skipping else-after-return, so would be written as:
```
  return quotient - 1;
return quotient;
```

lipracer commented:
> @lipracer could you open a separate PR for the APInt change? This will make it easier to review and land properly.

APInt change has already merged.

kuhar reviewed (approved):
Looks fine upon a quick scan but I won't have the time to understand the details of the expansion in near future. Please get a second approval before landing.

lipracer commented:
Thank you very much for your valuable suggestion.

lipracer commented:
@joker-eph  I'd like someone to review my code, could you help me with that?

Mogball reviewed (approved):
This change almost certainly has performance implications. Do we have some way to measure or even guess at that? If this is fixing a correctness bug, I am also not the right person to verify that, but I trust the author in this

joker-eph commented:
@pingshiyu  would you be able to review the correctness of this change?

lipracer commented:
> This change almost certainly has performance implications. Do we have some way to measure or even guess at that? If this is fixing a correctness bug, I am also not the right person to verify that, but I trust the author in this

Yes, this may indeed affect performance. This implementation is based on tensorflow. For performance considerations, I think we can have a compilation option similar to ‘fast-expand’ for expansion. This change is only consistent with the behavior of llvm.

pingshiyu commented:
> @pingshiyu would you be able to review the correctness of this change?

sure, I've just taken some time to check through the cases, lgtm re correctness. 

thanks @lipracer!

maybe some extra tests would be nice for the edge cases, e.g. `floordivsi(min_val, 1)`, and `floordivsi(max_val, -1)`

lipracer committed:
add more corner test

lipracer commented:
@pingshiyu Thanks, I have added more corner tests.

lipracer commented:
The CI error seems unrelated to this change. If the performance drops after merging, I will submit another change to enable "fast expand".

joker-eph commented:
The premerge is green at HEAD: can you rebase? https://lab.llvm.org/buildbot/#/builders/271

joker-eph commented:
Nevermind I misread the log, it's failing on building flang by running out of memory:
```
_bk;t=1711039712398C:\ws\src\flang\include\flang\Evaluate\tools.h(236): fatal error C1060: compiler is out of heap space
```

I concur that it is unrelated, and all the MLIR tests are passing.

lipracer commented:
Thanks, I will rebase on head.
